File: C/13.c
//lucianopinheirob

#include <stdio.h>
#define indice_max 6

int numeros_possiveis[] = {1000, 500, 100, 50, 10, 5, 1};
char letras_possiveis[] = {'M', 'D', 'C', 'L', 'X', 'V', 'I'};

int retorna_indice(char letra);
int romanToInt(char* S);

int main (){
    char* numero_romano;
    int numero_inteiro;
    numero_romano = "MCDXCIV";
    numero_inteiro = romanToInt(numero_romano);
    printf("%d", numero_inteiro);
    return 0;
}

int retorna_indice(char letra){
    int indice = 0;
    while (indice < indice_max){
            if (letra == letras_possiveis[indice]) break;
            indice += 1;
        }
    return indice;
}

int romanToInt (char* S){
    int numero_inteiro = 0, indice, proximo_indice, delta_indice;
    while (*S != '\0'){
        indice = retorna_indice(*S);
        proximo_indice = retorna_indice(*(S+1));
        delta_indice = indice - proximo_indice;
        if (delta_indice > 0){
            numero_inteiro += numeros_possiveis[proximo_indice] - numeros_possiveis[indice];
            S = S + 2;
        }
        else{
            numero_inteiro += numeros_possiveis[indice];
            S += 1;
        } 
    }
    return numero_inteiro;
}

File: Python/13.py
#lucianopinheirob
#Como um desafio a mais, esse código foi feito com um loop mais "pythonico".

class Solution:
    def romanToInt(self, s: str) -> int:
       
        dicio = {
            'M': 1000,
            'D': 500,
            'C': 100,
            'L': 50,
            'X': 10,
            'V': 5,
            'I': 1
            }
        
        indice_max = len(s) - 1
        numero_inteiro = 0
        ocorreu_excecao = False

        for indice, letra in enumerate(s):
            
            #Skipa um índice quando ocorre uma exceção
            if ocorreu_excecao:
                ocorreu_excecao = False
                continue
            
            #Checa se já chegou no final da string
            if indice != indice_max:
                valor_atual = dicio[s[indice]]
                valor_proximo = dicio[s[indice+1]]
                delta_valor = valor_atual - valor_proximo
            else:
                valor_atual = dicio[s[indice]]
                valor_proximo = 0
                delta_valor = valor_atual - valor_proximo
            
            #Trata exceção
            if delta_valor < 0:
                ocorreu_excecao = True
                numero_inteiro -= delta_valor
                continue
            
            numero_inteiro += valor_atual
            
        return numero_inteiro
            
solucao = Solution()
print(solucao.romanToInt('CMXIV'))
